finality-aleph/src/validator_network/manager/direction.rs

```diff
@@ -0,0 +1,212 @@
+use std::{
+    collections::{HashMap, HashSet},
+    ops::BitXor,
+};
+
+use aleph_primitives::AuthorityId;
+
+use crate::validator_network::Data;
+
+/// Data about peers we know and whether we should connect to them or they to us. For the former
+/// case also keeps the peers' addresses.
+pub struct DirectedPeers<A: Data> {
+    own_id: AuthorityId,
+    outgoing: HashMap<AuthorityId, Vec<A>>,
+    incoming: HashSet<AuthorityId>,
+}
+
+/// Whether we should call the remote or the other way around. We xor the peer ids and based on the
+/// parity of the sum of bits of the result decide whether the caller should be the smaller or
+/// greated lexicographically. They are never equal, because cryptography.
+fn should_we_call(own_id: &[u8], remote_id: &[u8]) -> bool {
```

Comment:
This looks like a custom implementation of equality operator on `AuthorityId`. I'm pretty sure there should exist a version of it that exits early, without doing the computation over whole thing.

At the same time, optimizing this seems totally unnecessary.

Reply:
Huh, equality? It's a relation, but definitely not equality.

Reply:
Why not? This method could be called for `own_id == remote_id` so all the outcomes are possible: less-than, more-than and equal.

Reply:
Ah, you're correct of course. I was talking about `Ord` but wrote about equality.

Reply:
It's also not `Ord`, because it's not transitive, and that is by design. Otherwise there would be a minimal element and someone could control this way whether nodes want to connect to them or they to nodes (to an extent, they cannot actually pick _the_ minimal element, but they could generate keys until they get a sufficiently small/big one). I don't know of any way this control could be abused, but I see no reason for providing it to attackers, since making this nontransitive isn't particularly hard.